### PR TITLE
Fix data race accessing shared slice from goroutines in dependency computation.

### LINKED ISSUE #2873: Error in call to function "get_repo_root" - "git" executable not found in $PATH failure
**Describe the bug**
We are currently experiencing intermittent build failures in CircleCI when running `validate` commands against multiple modules in parallel (our current configuration is `parallelism: 25`). The issue only surfaces on some of the parallelised `validate` jobs, and we're utilising CircleCI's test splitting functionality for this.

The error we're currently getting is below:

```
time=2023-12-28T19:57:53Z level=error msg=[module path omitted]/terragrunt.hcl:7,15-29: Error in function call; Call to function "get_repo_root" failed: exec: "git": executable file not found in $PATH., and 1 other diagnostic(s)
time=2023-12-28T19:57:53Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1
Makefile:28: recipe for target '<some target>' failed
```

We're running this in a Docker image, which definitely has `git` installed. A snippet of the Dockerfile is below:

```
ARG DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
      automake \
      autoconf \
      bsdmainutils \
      ca-certificates \
      libreadline-dev \
      libncurses-dev \
      libssl-dev \
      libyaml-dev \
      libxslt-dev \
      libffi-dev \
      libtool \
      tzdata \
      unixodbc-dev \
      unzip \
      curl \
      git && \
    rm -rf /var/lib/apt/lists/*

ENV PATH="$PATH:/root/.asdf/shims:/root/.asdf/bin"
<other asdf related things>
```

**To Reproduce**
- Build a docker image with the above tooling and a base image of `ubuntu:22.04`, additionally install `terragrunt` and `terraform` inside the image, we copy in a large number of arbitrarily nested modules inside the container
- Run the following through CircleCI
```
< ... snipped ... >
  validate:
    executor: default
    parallelism: 25
    steps:
    - setup
    - run:
        name: Build image
        command: <snipped>
    - run:
        name: Validate
        command: |
          find <basedir>/ -mindepth 2 -name "terragrunt.hcl" -printf "%h\n" | \
          circleci tests split | \
          xargs -I {} docker run --rm \
		-v /tmp/.terraform.d/plugin-cache:/root/.terraform.d/plugin-cache \
		-e DISABLE_BACKEND=true \
                 <img> terragrunt -w {} validate
< ... snipped ... >
```

If you need more of the CircleCI config I can probably redact most of it.

**Expected behavior**
I'd expect each of the parallelised jobs to be able to find the `git` executable, since it's only ever between 1-4 jobs that fail, the rest succeed and can find the binary fine.

**Nice to have**
- [ ] Terminal output
- [ ] Screenshots

**Versions**
- Terragrunt version: 0.54.1
- Terraform version: 1.6.3
- Environment details (Ubuntu 20.04, Windows 10, etc.): CircleCI `ubuntu-2204:2023.10.1` machine image, with `docker_layer_caching` enabled

**Additional context**
This issue only presented itself when we bumped to `>=0.50.0`. It should be mentioned that at the same we also went from Terraform version `1.4.5` --> `1.6.3`. It's hard to tell whether is issue is because of the plugin-cache volume mount (of which I have seen some issues), or this is an issue with Terragrunt itself.

Thanks in advance!

### PR DESCRIPTION
**Pull request by vladlosev**
## Description

Fixes #2873.

This PR adds synchronization around concurrent access from goroutines to a shared slice. The concurrent access was probabilistically resulting in data corruption that was causing the problem described in the issue or (more rarely) a crash with the following stack dump:
```
panic: no such attribute

goroutine 18 [running]:
github.com/zclconf/go-cty/cty.Type.AttributeOptional({{0x255e7f0?, 0xc000faaa60?}}, {0xc0005f2580?, 0x1c?})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/object_type.go:214 +0xb5
github.com/zclconf/go-cty/cty/convert.mismatchMessageObjects({{0x255e7f0?, 0xc000faaa60?}}, {{0x255e7f0?, 0xc000faaa60?}})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/convert/mismatch_msg.go:81 +0x571
github.com/zclconf/go-cty/cty/convert.MismatchMessage({{0x255e7f0?, 0xc000faaa60?}}, {{0x255e7f0?, 0xc000faaa60?}})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/convert/mismatch_msg.go:33 +0x3e
github.com/zclconf/go-cty/cty/convert.mismatchMessageObjects({{0x255e7f0?, 0xc000faaa70?}}, {{0x255e7f0?, 0xc000faaa70?}})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/convert/mismatch_msg.go:102 +0x367
github.com/zclconf/go-cty/cty/convert.MismatchMessage({{0x255e7f0?, 0xc000faaa70?}}, {{0x255e7f0?, 0xc000faaa70?}})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/convert/mismatch_msg.go:33 +0x3e
github.com/zclconf/go-cty/cty/convert.Convert({{{0x255e7f0?, 0xc000faaa70?}}, {0x1b017e0?, 0xc000c64960?}}, {{0x255e7f0?, 0xc000faaa70?}})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/convert/public.go:49 +0x16d
github.com/zclconf/go-cty/cty/gocty.toCtyPassthrough({0x1d5c120?, 0xc00084e040?, 0x1d5c120?}, {{0x255e7f0?, 0xc000faaa70?}}, {0xc0003240c0, 0x1, 0x1})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/gocty/in.go:521 +0x97
github.com/zclconf/go-cty/cty/gocty.toCtyValue({0x1d5c120?, 0xc00084e040?, 0x1d7b2db?}, {{0x255e7f0, 0xc000faaa70}}, {0xc0003240c0, 0x1, 0x1})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/gocty/in.go:39 +0x559
github.com/zclconf/go-cty/cty/gocty.toCtyObject({0x1b02b60?, 0xc000812120?, 0x1d5c120?}, 0xc000812180, {0xc0010e0e50, 0x0, 0x0})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/gocty/in.go:337 +0x9bd
github.com/zclconf/go-cty/cty/gocty.toCtyValue({0x1b02b60?, 0xc000812120?, 0xc000812120?}, {{0x255e7f0, 0xc0003240b0}}, {0xc0010e0e50, 0x0, 0x0})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/gocty/in.go:61 +0x36c
github.com/zclconf/go-cty/cty/gocty.ToCtyValue({0x1b02b60?, 0xc000812120?}, {{0x255e7f0?, 0xc0003240b0?}})
        /Users/vladimir.losev/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/gocty/in.go:32 +0x9c
github.com/gruntwork-io/terragrunt/config.dependencyBlocksToCtyValue.func1()
        /Users/vladimir.losev/dev/terragrunt/config/dependency.go:324 +0x229
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /Users/vladimir.losev/go/pkg/mod/golang.org/x/sync@v0.4.0/errgroup/errgroup.go:75 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
        /Users/vladimir.losev/go/pkg/mod/golang.org/x/sync@v0.4.0/errgroup/errgroup.go:72 +0x96
```

We were seeing the problem only on CircleCI running in a Docker container, and it was occurring in approximately 0.15-0.35% of all Terragrunt runs. But given heavy parallelization of our tests, it was affecting ~5-15% of our CI runs. With a patched copy of Terragrunt, we saw the issue go away.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Fixed data race in dependency computation.

### CODE REVIEW

File: config/dependency.go [hunk 1]
```diff
@@ -330,7 +330,9 @@ func dependencyBlocksToCtyValue(ctx *ParsingContext, dependencyConfigs []Depende
 				return err
 			}
 			if dependencyConfig.RenderedOutputs != nil {
+				lock.Lock()
```
Comment (denis256) [inline config/dependency.go:333]: I was wondering if should use a shared lock with `dependencyMap`

Comment (denis256) [reply config/dependency.go:333]: I mean that `lock := sync.Mutex{}` defined in 318 is used to isolate access to different structures: `paths` and `dependencyMap` 
https://github.com/gruntwork-io/terragrunt/blob/c9e00e17d2ee8ef65e7a8623a2f33d6d912d164a/config/dependency.go#L334

https://github.com/gruntwork-io/terragrunt/blob/c9e00e17d2ee8ef65e7a8623a2f33d6d912d164a/config/dependency.go#L352

Race condition will occur if will be used separated lock instances? one for `paths` and different for `dependencyMap` ?

**Comment by denis256** [comment]
```
[INFO] Initializing environment for https://github.com/gruntwork-io/pre-commit.
Terraform fmt............................................................Failed
- hook id: terraform-fmt
- files were modified by this hook

outputs.tf

goimports................................................................Passed

Exited with code exit status 1
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -121,6 +121,7 @@ jobs:
             mkdir -p logs
             go mod tidy
             run-go-tests --parallelism 1 --packages "$(go list ./... | grep -v /test | tr '\n' ' ')"  | tee logs/unit.log
+            go test -run=TestParseDependencyBlockMultiple -race ./config | tee -a logs/unit.log
```
Comment (denis256) [inline .circleci/config.yml:124]: :+1: will add additional ticket to extend race testing across multiple tests

**Review by denis256 — APPROVED** [review]

**Comment by juliusmarcus38** [comment]
issue persist while executing Terragrunt in docker container whereas `foo=yamldecode(file("${get_repo_root()}/bar.yaml}"))` fails with error `ERRO[0000] Invalid value for "path" parameter: no file exists at....` 

any idea how to resolve?

**Comment by vladlosev** [comment]
Is the issue you are seeing happens only intermittently?

**Pull request merged**